src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+        /// </summary>
+        /// <param name="chainedHeader">The header where we start walking back the chain from.</param>
+        /// <param name="networkPeerId">The peer id that is removed.</param>
+        private void RemovePeerClaim(int networkPeerId, ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:{1},{2}:'{3}')", nameof(networkPeerId), networkPeerId, nameof(chainedHeader), chainedHeader);
+
+            var listOfPeersClaimingThisHeader = this.peerIdsByTipHash.TryGet(chainedHeader.HashBlock);
+
+            if (listOfPeersClaimingThisHeader == null)
+            {
+                this.logger.LogTrace("(-)[PEER_TIP_NOT_FOUND]");
+                throw new ConsensusException("PEER_TIP_NOT_FOUND");
+            }
+
+            this.logger.LogTrace("Peer id {0 } removed.", networkPeerId);
+            listOfPeersClaimingThisHeader.Remove(networkPeerId);
+
+            if (listOfPeersClaimingThisHeader.Count == 0)
+            {
+                this.logger.LogTrace("Header '{0}' is not the tip of any peer.", chainedHeader);
+                this.peerIdsByTipHash.Remove(chainedHeader.HashBlock);
+                this.RemoveUnclaimedBranch(chainedHeader);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Set a new header as a tip for this peer and remove the old tip.
+        /// </summary>
+        /// <param name="networkPeerId">The peer id that sets a new tip.</param>
+        /// <param name="newTip">The new tip to set.</param>
+        private void SetPeerTip(int networkPeerId, uint256 newTip)
+        {
+            this.logger.LogTrace("({0}:{1},{2}:'{3}')", nameof(networkPeerId), networkPeerId, nameof(newTip), newTip);
+
+            uint256 oldTipHash = this.peerTipsByPeerId.TryGet(networkPeerId);
+
+            var listOfPeersClaimingThisHeader = this.peerIdsByTipHash.TryGet(newTip);
+            if (listOfPeersClaimingThisHeader == null)
+            {
+                listOfPeersClaimingThisHeader = new HashSet<int>();
+                this.peerIdsByTipHash.Add(newTip, listOfPeersClaimingThisHeader);
+            }
+
+            listOfPeersClaimingThisHeader.Add(networkPeerId);
+            this.peerTipsByPeerId.AddOrReplace(networkPeerId, newTip);
+
+            if (oldTipHash != null)
+            {
+                ChainedHeader oldTip = this.chainedHeadersByHash.TryGet(oldTipHash);
+                this.RemovePeerClaim(networkPeerId, oldTip);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Find the headers that are not part of the tree and try to connect them to an existing chain 
+        /// by creating a new <see cref="ChainedHeader"/> type and linking it to its previous header.
+        /// </summary>
+        /// <remarks>
+        /// A new header will perform partial validation.
```

Comment:
It is not the header that performs the validation and it is not partial validation as talk about it elsewhere.

`Header validation is performed on each header`